src/mlpack/methods/reinforcement_learning/async_learning_impl.hpp

```diff
@@ -0,0 +1,108 @@
+/**
+ * @file async_learning_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * This file is the implementation of AsyncLearning class,
+ * which is wrapper for various asynchronous learning algorithms.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_ASYNC_LEARNING_IMPL_HPP
+#define MLPACK_METHODS_RL_ASYNC_LEARNING_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace rl {
+
+template <
+  typename WorkerType,
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType
+>
+AsyncLearning<
+  WorkerType,
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType
+>::AsyncLearning(TrainingConfig config,
+                 NetworkType network,
+                 PolicyType policy,
+                 UpdaterType updater,
+                 EnvironmentType environment):
+  config(std::move(config)),
+  learningNetwork(std::move(network)),
+  policy(std::move(policy)),
+  updater(std::move(updater)),
+  environment(std::move(environment))
+{ /* Nothing to do here. */ };
+
+template <
+  typename WorkerType,
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType
+>
+template <typename Measure>
+void AsyncLearning<
+  WorkerType,
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType
+>::Train(Measure& measure)
+{
+  /**
+   * OpenMP doesn't support shared class member variables.
+   * So we need to copy them to local variables.
+   */
+  NetworkType learningNetwork = std::move(this->learningNetwork);
+  if (learningNetwork.Parameters().is_empty())
+    learningNetwork.ResetParameters();
+  NetworkType targetNetwork = learningNetwork;
+  bool stop = false;
+  size_t totalSteps = 0;
+  PolicyType policy = this->policy;
+
+  // Make sure OpenMP will start desired number of threads.
+  omp_set_dynamic(0);
+  // An extra thread for test.
+  omp_set_num_threads(config.NumOfWorkers() + 1);
```

Comment:
I'm sorry to throw this comment in so late.  My understanding is that ideally `NumOfWorkers()` needs to be large for the algorithm to work.  But the code like this doesn't allow a user to limit the number of cores that the program uses, which is I think an important use case in resource-limited environments or other environments where there may be some contention.

But I see that it is important that there be enough workers, otherwise we will not get updates for each worker at each time step.  I think it might be better to adjust the design so that `Episode` doesn't run all iterations at once, but instead runs a single iteration at a time.  i.e., the loop here would look like:

```
for (size_t i = 0; i < numIterations; ++i)
{
  #pragma omp for ...
  for (omp_size_t j = 0; j < config.NumOfWorkers(); ++i)
  {
    ...
  }
}
```

and in this way you can get an update from each iteration of each episode even when having `OMP_NUM_THREADS` set to fewer than the number of workers.  If a problem with that approach is that each worker could take a different amount of time per iteration, you could adapt the idea above to use OpenMP tasks too.